### PR TITLE
Update Password Policy

### PR DESCRIPTION
**Pull request by theck13**
### Fix
Update the password policy requirements to the following:
- Password cannot match username
- Minimum of 8 characters
- No new lines
- No tabs

### Test
These changes are best tested on Simplenote by pointing your local Simplenote repository to your local Simperium repository.  Contact me for details.  In order to test the login password length updates, change your Simplenote password to more than four and less than eight characters.  In the Simplenote app, follow the steps below.

0. Clear app data.
1. Tap ***Sign Up*** button.
2. Enter email address in ***Email*** field.
3. Notice ***Sign Up*** button is disabled.
4. Enter email address, new lines, or tabs in ***Password*** field.
5. Notice ***Sign Up*** button is enabled after eight characters are input.
6. Tap ***Sign Up*** button.
7. Notice ***Error*** dialog with password requirements is shown.
8. Tap ***OK*** button in dialog.
9. Tap back arrow in top app bar.
10. Tap ***Log In*** button.
11. Tap ***Log In with Email*** button.
12. Enter email address in ***Email*** field.
13. Notice ***Log In*** button is disabled.
14. Enter password of more than four and less than eight characters in ***Password*** field.
15. Notice ***Log In*** button is enabled after four characters are input.
16. Tap ***Log In*** button.
17. Notice ***Error*** dialog with password requirements is shown.
18. Tap ***Reset*** button in dialog.
19. Notice external browser is opened showing the https://app.simplenote.com/forgot/ page.

### Review
Only one developer is required to review these changes, but anyone can perform the review.

### CODE REVIEW

File: Simperium/src/main/java/com/simperium/android/CredentialsActivity.java
Comment (mchowning) [inline Simperium/src/main/java/com/simperium/android/CredentialsActivity.java:314]: Should this be getting checked in the `setButtonState()` method? The sign up button is enabled for me when I put the same email address in the password field.

File: Simperium/src/main/java/com/simperium/android/CredentialsActivity.java
Comment (mchowning) [inline Simperium/src/main/java/com/simperium/android/CredentialsActivity.java]: #### Allowing _some_ \n and \t?

I believe `PATTERN_EXCLUDE_NEWLINES_TABS.matcher(password).find()` (aka `Pattern.compile("[^\n\t]+").matcher(password).find()`) will return true if the pasword contains at least one character that is not a newline or a tab (i.e., `\n\n\n\n\n\n\na` returns true). The PR description and the error dialog both make it sound like we don't want _any_ newlines or tabs. I'm not sure which is right.

#### Spaces?

The PR description and the error message don't mention spaces, so I assume those are allowed now? For example, a password like "a\s\s\s\s\s\s\sa" is now valid? Just wanted to confirm. Interestingly, using a password like that seemed to work on the web, but when I try to log in using it on mobile I'm getting a "Could not log in with the provided email address and password" error.

#### Unit tests?
At a higher level, it seems like it could be useful to pull the password validation logic to a separate class. That would make it easy to write some unit tests with example valid/invalid inputs. Tests like that could obviously help catch bugs, but more importantly imo they would make the intention behind the code really clear (I love examples 🙂). Just my $.02, not something that needs to block the PR if you disagree.

File: Simperium/src/main/res/values/strings.xml
Comment (mchowning) [inline Simperium/src/main/res/values/strings.xml:16]: This string resource seems to get overriden by a matching string resource in Simplenote: https://github.com/Automattic/simplenote-android/blob/develop/Simplenote/src/main/res/values/strings.xml#L239. I'm guessing that just hasn't been updated yet, but I wanted to point it out. Took me a bit to figure out why this message wasn't showing up in the dialog within Simplenote. 

FWIW, I think this dialog message is much clearer than the matching one in Simplenote. I also noticed that the one in Simplenote says that spaces are not allowed, which is different from this error message and the implementation in this PR.

Another small detail is that I noticed that this error message says the password cannot match the "username", but there is no username field, just an email field. It's pretty obvious what is meant, and I have trouble imagining many people getting confused by that, but figured I'd point it out.

File: Simperium/src/main/java/com/simperium/android/CredentialsActivity.java
Comment (mchowning) [inline Simperium/src/main/java/com/simperium/android/CredentialsActivity.java]: We aren't doing the regex check here for newlines and tabs, so that means that although newlines or tabs cannot be used when signing up, we will not force someone with them in their password to reset their password, right? I mainly wanted to make sure we meant to have different logic here and on [line 425](https://github.com/Simperium/simperium-android/pull/219/files#diff-f578252d8746b4d32cf359ef222428b4R425). 

I'm also wondering because these kinds of passwords don't seem to work on the web, even though I could create a password with tabs on the web. When I set up a password through app.simplenote.com, I was allowed to set "test\t\t1234" as my password. The mobile app would also allow me to use it (i.e., it wouldn't force me to reset it), but it would then get rejected by the server. Interestingly, this password was also (sort of) rejected in the web app (where I set up the password) because using it to log in just drops me back to the login screen without any error message (such as the one that shows up when I enter the wrong password). 

This gif can be a little hard to follow so I tried to add some context in the email field. Basically, it shows me using app.simplenote.com and first logging in with the correct "test\t\t1234" password and getting a black screen before getting dropped back to the same login page I was on before. Then I login using an incorrect password and I get the expected error message. Basically, I'm not sure I can log into that account anymore now that I changed the password to "test\t\t1234". It is worth noting though that about 1 out of every 5 or 6 times I try to log in with the "test/t/t1234" password I do get a captcha failed error message (even though I never saw any captcha).

password_problems

**Review by mchowning — COMMENTED** [review]
This is nice @theck13 ! 👍 I did notice a couple of issues, but they may just need some clarification more than code changes. 

Also, I was observing some surprising behavior/inconsistencies in the password handling by the server on mobile and app.simplenote.com. The behavior was weird enough that I would certainly not rule out me just having something weird locally, so if any of my comments seem unlikely please don't hesitate to reach out and have me double check. 🙂

File: Simperium/src/main/res/values/strings.xml
Comment (theck13) [reply Simperium/src/main/res/values/strings.xml:16]: > I'm guessing that just hasn't been updated yet

Exactly!  We have to update the `simperium_dialog_message_password` resource in Simplenote once this pull request is merged, the Simperium library is published, and the Simperium dependency is updated in Simplenote.  There will be a subsequent pull request in Simplenote to do just that.

> Another small detail is that I noticed that this error message says the password cannot match the "username", but there is no username field, just an email field.

That's true for Simplenote that the username is the email address.  For the Simperium library, I wanted to keep it generic and use username since it _could_ be used in an app other than Simplenote.  Fortunately, we can just override `simperium_dialog_message_password` in Simplenote to fix that.

Thanks for pointing out those things!

File: Simperium/src/main/java/com/simperium/android/CredentialsActivity.java
Comment (theck13) [reply Simperium/src/main/java/com/simperium/android/CredentialsActivity.java]: > I believe `PATTERN_EXCLUDE_NEWLINES_TABS.matcher(password).find()` (aka `Pattern.compile("[^\n\t]+").matcher(password).find())` will return true if the pasword contains at least one character that is not a newline or a tab

You're right!  My mistake.  I updated the regular expression and negated that part of the return condition of `isValidPassword` in ffc5f1c.

> The PR description and the error message don't mention spaces, so I assume those are allowed now?

Yes.  The original pattern (i.e. `Pattern.compile("(\\s)")`) matched all whitespace including spaces and didn't allow any.  It was decided to allow spaces, but not new lines or tabs.  That's why you can't use spaces in the current Simplenote Android app, but you can on the web.

> That would make it easy to write some unit tests with example valid/invalid inputs.

I agree.  Unfortunately, tests have no been a focus for Simperium or Simplenote for a long time.  That could be its own project just to get a decent amount of coverage for both.  I'll bring it up with the team.

File: Simperium/src/main/java/com/simperium/android/CredentialsActivity.java
Comment (theck13) [reply Simperium/src/main/java/com/simperium/android/CredentialsActivity.java]: > We aren't doing the regex check here for newlines and tabs, so that means that although newlines or tabs cannot be used when signing up, we will not force someone with them in their password to reset their password, right?

You're right.  The current code does not force them to reset their password if `isValidPassword` returns `false`, but we should.  I updated that `if`/`else` statement in ee17a06.

This is the first platform to enforce the new password policy.  That's why you're seeing some odd behavior with the Simplenote web app.  The web and others will be updated soon.

File: Simperium/src/main/java/com/simperium/android/CredentialsActivity.java
Comment (mchowning) [inline Simperium/src/main/java/com/simperium/android/CredentialsActivity.java]: We're now checking `isValidPassword(password) && !isEmailPasswordMatch()` both here and on [line 423](https://github.com/Simperium/simperium-android/pull/219/files#diff-f578252d8746b4d32cf359ef222428b4R423). Three questions about that:

1. We're [trimming the password](https://github.com/Simperium/simperium-android/pull/219/files#diff-f578252d8746b4d32cf359ef222428b4R406), which allows a user to add invalid characters (\n\t) to the end of their password, and a different password (with those invalid characters "trimmed") gets accepted and saved. Similarly, the user could end their password with now-valid space characters, and those spaces would get removed without the user's knowledge. Admittedly, it seems like this kind of works with our current Android implementation because we also trim the password on login, so the user can submit the same password with invalid characters that they signed up with, and it works because we again trim it before sending it to the server. We would have to make sure that trim logic is used on every platform to avoid issues though. It's hard for me to imagine that won't cause us problems at some point.

2. Although the password that gets saved is being trimmed here, we're not trimming it [when we're checking for an email password match inside the `isEmailPasswordMatch` method](https://github.com/Simperium/simperium-android/pull/219/files#diff-f578252d8746b4d32cf359ef222428b4R315-R316). This means that an email/password of "me@gmail.com"/"me@gmail.com " (note the space at the end of the password) is accepted as valid because the untrimmed password does not match the email. But the trimmed password is saved, and the trimmed password is the same as the email, which seems like that is what we're trying to stop with the `isEmailPasswordMatch()` check.

3. What would you think about moving the `!isEmailPasswordMatch()` check inside the `isValidPassword(...)` function? We do the email match check always (and only) when we're checking `isValidPassword(...)`, so it would remove some duplication. More importantly, I feel like it makes sense from a logic point of view since the email check is part of checking if a password is valid. I would even consider removing the `isEmailPasswordMatch()` method and just assigning the logic to a boolean variable inside `isValidPassword(...)` because that would make it relatively easy to catch inconsistencies like (2) above.

Comment (theck13) [reply Simperium/src/main/java/com/simperium/android/CredentialsActivity.java]: We shouldn't be trimming the `email` and `password` variables anymore since we do allow spaces.  I removed trimming in f63c491.

I moved the logic from `isEmailPasswordMatch` to `isValidPassword` in e29fd60.

File: Simperium/src/main/java/com/simperium/android/CredentialsActivity.java
Comment (mchowning) [inline Simperium/src/main/java/com/simperium/android/CredentialsActivity.java]: I think it would be better to use the `password` parameter that is passed into this function for the email check instead of re-retrieving it from the input field.

Similarly, every time we call this method we have [already retrieved the email value from the input field](https://github.com/Simperium/simperium-android/pull/219/files#diff-f578252d8746b4d32cf359ef222428b4R407), if we passed that email value into this method instead of re-retrieving it here inside the method that would insure that we're using the same "email" in both places and simplify the logic here.

Comment (theck13) [reply Simperium/src/main/java/com/simperium/android/CredentialsActivity.java]: Good idea!  I added the `email` parameter to `isValidPassword` in ef59a99.

**Review by mchowning — APPROVED** [review]
Working great! Thanks Tyler! 🎉

**Pull request merged**
